File: Semestrul 4/MPP2/Client-Server_2Limbaje/Problema8SC_CSharp/Networking/jsonprotocol/ClientJsonWorker.cs
using Networking.dto;

namespace Networking.jsonprotocol;

using System;
using System.IO;
using System.Net.Sockets;
using System.Text;
using System.Threading;
using System.Text.Json;
using System.Collections.Generic;
using log4net;

using Problema8SC_CSharp.Model;
using Problema8SC_CSharp.Services;

public class ClientJsonWorker : IObserver
{
    private IServices server;
    private TcpClient connection;
    private NetworkStream stream;
    private volatile bool connected;
    private static readonly ILog log = LogManager.GetLogger(typeof(ClientJsonWorker));

    private static Response okResponse = JsonProtocolUtils.CreateOkResponse();

    public ClientJsonWorker(IServices server, TcpClient connection)
    {
        this.server = server;
        this.connection = connection;
        try
        {
            stream = connection.GetStream();
            connected = true;
        }
        catch (Exception e)
        {
            log.Error(e.StackTrace);
        }
    }

    public void Run()
    {
        using StreamReader reader = new StreamReader(stream, Encoding.UTF8);
        while (connected)
        {
            try
            {
                string requestJson = reader.ReadLine();
                if (string.IsNullOrEmpty(requestJson))
                {
                    continue;
                }
                log.Debug($"Received JSON request: {requestJson}");
                Request request = JsonSerializer.Deserialize<Request>(requestJson);
                Response response = HandleRequest(request);
                if (response != null)
                {
                    SendResponse(response);
                }
            }
            catch (Exception e)
            {
                log.Error($"Run error: {e.Message}");
                if (e.InnerException != null)
                    log.Error($"Inner error: {e.InnerException.Message}");
                log.Error(e.StackTrace);
                connected = false;
            }

            try
            {
                Thread.Sleep(1000);
            }
            catch (Exception e)
            {
                log.Error(e.StackTrace);
            }
        }

        try
        {
            stream.Close();
            connection.Close();
        }
        catch (Exception e)
        {
            log.Error($"Closing connection error: {e.Message}");
        }
    }

    private Response HandleRequest(Request request)
    {
        Response response = null;
        if (request.Type == 2)
        {
            log.Debug("Login request...");
            User user = DTOUtils.GetFromDTO(request.User);
            try
            {
                lock (server)
                {
                    server.Login(user, this);
                }
                return okResponse;
            }
            catch (ProbExceptions e)
            {
                connected = false;
                return JsonProtocolUtils.CreateErrorResponse(e.Message);
            }
        }
        if (request.Type == 3)
        {
            log.Debug("Logout request...");
            User user = DTOUtils.GetFromDTO(request.User);
            try
            {
                lock (server)
                {
                    server.Logout(user, this);
                }
                connected = false;
                return okResponse;
            }
            catch (ProbExceptions e)
            {
                connected = false;
                return JsonProtocolUtils.CreateErrorResponse(e.Message);
            }
        }
        if (request.Type == 4)
        {
            log.Debug("Get age groups request...");
            try
            {
                var groups = server.GetAgeGroups();
                return JsonProtocolUtils.CreateGetAgeGroupsResponse(groups);
            }
            catch (ProbExceptions e)
            {
                connected = false;
                return JsonProtocolUtils.CreateErrorResponse(e.Message);
            }
        }
        if (request.Type == 5)
        {
            log.Debug("Get events request...");
            try
            {
                var events = server.GetEvents();
                return JsonProtocolUtils.CreateGetEventsResponse(events);
            }
            catch (ProbExceptions e)
            {
                connected = false;
                return JsonProtocolUtils.CreateErrorResponse(e.Message);
            }
        }
        if (request.Type == 6)
        {
            log.Debug("Get events by age group request...");
            try
            {
                var events = server.GetEventsByAgeGroup(request.AgeGroupId);
                return JsonProtocolUtils.CreateGetEventsByAgeGroupResponse(events);
            }
            catch (ProbExceptions e)
            {
                connected = false;
                return JsonProtocolUtils.CreateErrorResponse(e.Message);
            }
        }
        if (request.Type == 7)
        {
            log.Debug("Save child and enrollment request...");
            try
            {
                server.SaveChildAndEnrollment(request.Nume, request.Cnp, request.EventName1);
                Child child = new Child(0, request.Nume, request.Cnp);
                Enrollment enrollment = new Enrollment(0, child, request.EventName1);
                
                return JsonProtocolUtils.CreateSaveChildAndEnrollmentResponse(child, enrollment);
            }
            catch (ProbExceptions e)
            {
                connected = false;
                return JsonProtocolUtils.CreateErrorResponse(e.Message);
            }
        }
        if (request.Type == 8)
        {
            log.Debug("Get children by event request...");
            try
            {
                var children = server.GetChildrenByEvent(request.Id);
                return JsonProtocolUtils.CreateGetChildrenByEventResponse(children);
            }
            catch (ProbExceptions e)
            {
                connected = false;
                return JsonProtocolUtils.CreateErrorResponse(e.Message);
            }
        }
        if (request.Type == 1)
        {
            log.Debug("Get children request...");
            try
            {
                var children = server.GetChildrens();
                return JsonProtocolUtils.CreateGetChildrenResponse(children);
            }
            catch (ProbExceptions e)
            {
                connected = false;
                return JsonProtocolUtils.CreateErrorResponse(e.Message);
            }
        }
        if (request.Type == 0)
        {
            log.Debug("Get details for all children request...");
            try
            {
                var children = server.GetDetailsForAllChildren();
                return JsonProtocolUtils.CreateGetDetailsForAllChildrenResponse(children);
            }
            catch (ProbExceptions e)
            {
                connected = false;
                return JsonProtocolUtils.CreateErrorResponse(e.Message);
            }
        }

        return response;
    }

    private void SendResponse(Response response)
    {
        string jsonString = JsonSerializer.Serialize(response);
        log.Debug($"Sending response: {jsonString}");
        lock (stream)
        {
            byte[] data = Encoding.UTF8.GetBytes(jsonString + "\n"); // newline to separate messages
            stream.Write(data, 0, data.Length);
            stream.Flush();
        }
    }
    
    public void UserLoggedIn(User user)
    {
        log.Debug($"User logged in: {user}");
        try
        {
            SendResponse(JsonProtocolUtils.CreateUserLoggedInResponse(user));
        }
        catch (Exception e)
        {
            log.Error($"Error notifying user logged in: {e.Message}");
        }
    }

    public void UserLoggedOut(User user)
    {
        log.Debug($"User logged out: {user}");
        try
        {
            SendResponse(JsonProtocolUtils.CreateUserLoggedOutResponse(user));
        }
        catch (Exception e)
        {
            log.Error($"Error notifying user logged out: {e.Message}");
        }
    }

    public void GetChildrenByEvent(List<Child> childrens)
    {
        log.Debug($"Update children by event: {childrens}");
        try
        {
            SendResponse(JsonProtocolUtils.CreateGetChildrenByEventResponse(childrens));
        }
        catch (Exception e)
        {
            log.Error($"Error sending children list: {e.Message}");
        }
    }

    public void SaveChildAndEnrollment(Child child, Enrollment enrollment)
    {
        log.Debug($"Child enrolled: {child}");
        try
        {
            SendResponse(JsonProtocolUtils.CreateChildEnrolledResponse());
        }
        catch (Exception e)
        {
            log.Error($"Error sending enrollment: {e.Message}");
        }
    }
}

File: Semestrul 4/MPP2/Client-Server_2Limbaje/Problema8SC/Problema8SC/Problema8Networking/src/main/java/problema8/network/jsonprotocol/ClientJsonWorker.java
package problema8.network.jsonprotocol;

import problema8.model.AgeGroup;
import problema8.model.Child;
import problema8.model.Enrollment;
import problema8.model.User;
import problema8.network.dto.AgeGroupDTO;
import problema8.network.dto.DTOUtils;
import problema8.network.dto.UserDTO;
import problema8.services.IObserver;
import problema8.services.IService;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.PrintWriter;
import java.net.Socket;
import java.util.List;

import com.google.gson.Gson;
import problema8.services.ProbExceptions;

public class ClientJsonWorker implements Runnable, IObserver {
    private IService server;
    private Socket connection;

    private BufferedReader input;
    private PrintWriter output;
    private Gson gsonFormatter;
    private volatile boolean connected;

    private static Logger logger = LogManager.getLogger(ClientJsonWorker.class);

    public ClientJsonWorker(IService server, Socket connection) {
        this.server = server;
        this.connection = connection;
        gsonFormatter = new Gson();
        try {
            output = new PrintWriter(connection.getOutputStream());
            input = new BufferedReader(new InputStreamReader(connection.getInputStream()));
            connected = true;
        } catch (IOException e) {
            logger.error(e);
            logger.error(e.getStackTrace());
        }
    }

    public void run() {
        while (connected) {
            try {
                String requestLine = input.readLine();
                if (requestLine == null) {
                    logger.info("Connection closed by client.");
                    connected = false;
                    break;
                }
                Request request = gsonFormatter.fromJson(requestLine, Request.class);
                Response response = handleRequest(request);
                if (response != null) {
                    sendResponse(response);
                }
            } catch (IOException e) {
                logger.error(e);
                logger.error(e.getStackTrace());
                connected = false;
            }
            try {
                Thread.sleep(1000);
            } catch (InterruptedException e) {
                logger.error(e);
                logger.error(e.getStackTrace());
            }
        }
        try {
            input.close();
            output.close();
            connection.close();
        } catch (IOException e) {
            logger.error("Error " + e);
        }
    }


    private void sendResponse(Response response) {
        String responseLine = gsonFormatter.toJson(response);
        logger.debug("Sending response: " + responseLine);
        synchronized (output) {
            output.println(responseLine);
            output.flush();
        }
    }

    private static Response okResponse=JsonProtocolUtils.createOkResponse();

    private Response handleRequest(Request request) {
        Response response = null;
        if(request.getType() == 2) {
            logger.debug("Login request...{}" + request.getUser());
            UserDTO udto = request.getUser();
            User user = DTOUtils.getFromDTO(udto);
            try {
                server.login(user, this);
                return okResponse;
            } catch (Exception e) {
                connected = false;
                return JsonProtocolUtils.createErrorResponse(e.getMessage());
            }
        }
        if(request.getType() == 3) {
            logger.debug("Logout request...{}" + request.getUser());
            UserDTO udto = request.getUser();
            User user = DTOUtils.getFromDTO(udto);
            try {
                server.logout(user, this);
                connected = false;
                return okResponse;
            } catch (Exception e) {
                connected = false;
                return JsonProtocolUtils.createErrorResponse(e.getMessage());
            }
        }
        if(request.getType() == 4) {
            logger.debug("Get age groups request...");
            try {
                response = JsonProtocolUtils.createGetAgeGroupsResponse(server.getAgeGroups());
                return response;
            } catch (ProbExceptions e) {
                connected = false;
                return JsonProtocolUtils.createErrorResponse(e.getMessage());
            }
        }
        if(request.getType() == 5) {
            logger.debug("Get events request...");
            try {
                response = JsonProtocolUtils.createGetEventsResponse(server.getEvents());
                return response;
            } catch (ProbExceptions e) {
                connected = false;
                return JsonProtocolUtils.createErrorResponse(e.getMessage());
            }
        }
        if(request.getType() == 6) {
            logger.debug("Get events by age group request...");
            try {
                response = JsonProtocolUtils.createGetEventsByAgeGroupResponse(server.getEventsByAgeGroup(request.getAge_group_id()));
                return response;
            } catch (ProbExceptions e) {
                connected = false;
                return JsonProtocolUtils.createErrorResponse(e.getMessage());
            }
        }
        if(request.getType() == 7) {
            logger.debug("Save child and enrollment request...");
            try {
                server.saveChildAndEnrollment(request.getNume(), request.getCnp(), request.getEventName1());
                Child child = new Child(null, request.getNume(), request.getCnp());
                Enrollment enrollment = new Enrollment(null, child, request.getEventName1());
                response = JsonProtocolUtils.createSaveChildAndEnrollmentResponse(child, enrollment);
                return response;
            } catch (ProbExceptions e) {
                connected = false;
                return JsonProtocolUtils.createErrorResponse(e.getMessage());
            }
        }
        if(request.getType() == 8) {
            logger.debug("Get children by event request...");
            try {
                response = JsonProtocolUtils.createGetChildrenByEventResponse(server.getChildrenByEvent(request.getId()));
                return response;
            } catch (ProbExceptions e) {
                connected = false;
                return JsonProtocolUtils.createErrorResponse(e.getMessage());
            }
        }
        if(request.getType() == 1) {
            logger.debug("Get children request...");
            try {
                response = JsonProtocolUtils.createGetChildrenResponse(server.getChildren());
                return response;
            } catch (ProbExceptions e) {
                connected = false;
                return JsonProtocolUtils.createErrorResponse(e.getMessage());
            }
        }
        if(request.getType() == 9) {
            logger.debug("Get number of events request...");
            try {
                long id = request.getId();
                response = JsonProtocolUtils.createGetNumberOfEventsResponse(server.getNumberOfEvents(id));
                return response;
            } catch (ProbExceptions e) {
                connected = false;
                return JsonProtocolUtils.createErrorResponse(e.getMessage());
            }
        }
        if (request.getType() == 0) {
            logger.debug("Send details for all children request...");
            try {
                response = JsonProtocolUtils.createGetDetailsForAllChildrenResponse(server.getDetailsForAllChildren());
                return response;
            } catch (ProbExceptions e) {
                connected = false;
                return JsonProtocolUtils.createErrorResponse(e.getMessage());
            }
        }
        return response;
    }

    @Override
    public void userLoggedIn(User user) throws ProbExceptions {
        logger.debug("User logged in: " + user);
        Response response = JsonProtocolUtils.createUserLoggedInResponse(user);
        sendResponse(response);
    }

    @Override
    public void userLoggedOut(User user) throws ProbExceptions {
        logger.debug("User logged out: " + user);
        Response response = JsonProtocolUtils.createUserLoggedOutResponse(user);
        sendResponse(response);
    }

    @Override
    public void getChildrenByEvent(List<Child> childrens) throws ProbExceptions {
        logger.debug("Update children by event: " + childrens);
        Response response = JsonProtocolUtils.createGetChildrenByEventResponse(childrens);
        sendResponse(response);
    }

    @Override
    public void saveChildAndEnrollment(Child child, Enrollment enrollment) throws ProbExceptions {
        logger.debug("ClientJsonWorker - Child enrolled: " + child);
        Response response = JsonProtocolUtils.createChildEnrolledResponse();
        sendResponse(response);
    }
}
